api/reauthenticate.go

```diff
@@ -0,0 +1,98 @@
+package api
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"github.com/netlify/gotrue/api/sms_provider"
+	"github.com/netlify/gotrue/conf"
+	"github.com/netlify/gotrue/models"
+	"github.com/netlify/gotrue/storage"
+)
+
+// Reauthenticate sends a reauthentication otp to either the user's email or phone
+func (a *API) Reauthenticate(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	config := a.getConfig(ctx)
+	instanceID := getInstanceID(ctx)
+
+	claims := getClaims(ctx)
+	userID, err := uuid.FromString(claims.Subject)
+	if err != nil {
+		return badRequestError("Could not read User ID claim")
+	}
+	user, err := models.FindUserByID(a.db, userID)
+	if err != nil {
+		if models.IsNotFoundError(err) {
+			return notFoundError(err.Error())
+		}
+		return internalServerError("Database error finding user").WithInternalError(err)
+	}
+
+	email, phone := user.GetEmail(), user.GetPhone()
+
+	if email == "" && phone == "" {
+		return unprocessableEntityError("Reauthentication requires the user to have an email or a phone number")
+	}
+
+	if email != "" {
+		if !user.IsConfirmed() {
```

Comment:
Could be combined into a single condition?
```
if email != "" &&  !user.IsConfirmed() {
}
```

Not critical though.

Reply:
hmm not really because it will fall through to the next case if the the user has an email and the user is confirmed which is not what we want